### PR TITLE
proxy: added endpoint refresh and timeout configuration values

### PR DESCRIPTION
**Pull request by webner**
the default dial timeout was set to 30 seconds this made the proxy a pain to use
in failure scenarios.

fixes 2862

### CODE REVIEW

**Comment by xiang90** [comment]
@webner I am OK with making these configurable. But probably we should not give a default timeout for `read`. We have long pulling watch request.

**Comment by webner** [comment]
good point. I removed the default read timeout (set it to 0) and added comment in the documentation that changing the read timeout will break watches.

File: Documentation/configuration.md [hunk 1]
```diff
@@ -105,6 +105,27 @@ To start etcd automatically using custom settings at startup in Linux, using a [
...
+
+##### -proxy-refresh-interval
++ Time (in milliseconds) of the endpoints refresh interval.
++ default: 30000
+
+##### -proxy-dial-timeout
++ Time (in milliseconds) for a dial to timeout.
++ default: 1000
+
+##### -proxy-write-timeout
```
Comment (xiang90) [inline Documentation/configuration.md]: can we also comment that 0 is no timeout?

File: Documentation/configuration.md [hunk 2]
```diff
@@ -105,6 +105,27 @@ To start etcd automatically using custom settings at startup in Linux, using a [
...
++ Time (in milliseconds) for a write to timeout.
++ default: 5000
+
+##### -proxy-read-timeout
++ Time (in milliseconds) for a read to timeout.
++ Don't change this value if you use watches because they are using long polling requests.
```
Comment (xiang90) [inline Documentation/configuration.md]: same here :)

File: proxy/director.go [hunk 3]
```diff
@@ -22,24 +22,17 @@ import (
 	"time"
 )
 
-const (
-	// amount of time an endpoint will be held in a failed
-	// state before being reconsidered for proxied requests
```
Comment (xiang90) [inline proxy/director.go]: I think we should still keep the second sentence. (move the whole comment to configuration.md).

File: proxy/director.go [hunk 4]
```diff
@@ -22,24 +22,17 @@ import (
...
-	refreshEndpoints = 30 * time.Second
-)
-
-func newDirector(urlsFunc GetProxyURLs) *director {
+func newDirector(urlsFunc GetProxyURLs, failureWait time.Duration, refreshInterval time.Duration) *director {
 	d := &director{
-		uf: urlsFunc,
+		uf:              urlsFunc,
+		failureWait:     failureWait,
+		refreshInterval: refreshInterval,
```
Comment (xiang90) [inline proxy/director.go]: Where is `refreshInterval` in director struct got used? Can we remove it if it is not used anywhere?

**Comment by xiang90** [comment]
@webner This change LGTM.

Thanks for the contribution! Have you tested the default configuration? I hope we can test it for a while in a high load env (have both watch, read, write) before merging it.

**Comment by webner** [comment]
I'm still testing the default configuration in our development cluster. So far it's looking good. Restarting a etcd node in our 3-node etcd cluster has now less impact on the worker nodes which are using an etcd proxy.

**Comment by xiang90** [comment]
Fix #2862 (fix the original reference)

**Comment by xiang90** [comment]
@webner Let me know when this is ready

**Comment by xiang90** [comment]
@webner Any update?

**Comment by webner** [comment]
Looks good so far. The proxy is running stable in our development cluster (~20 machines) and I had no issues with the default settings.

**Comment by xiang90** [comment]
@webner gofmt checks failed. Can you gofmt the proxy/direct.go code?

**Pull request merged**